src/tree.rs

```diff
@@ -1,15 +1,298 @@
+
+    domain
+        .hash(
+            iter::empty()
+                .chain(i2lebsp_k(l_star).iter().copied())
+                .chain(
+                    pair.left
+                        .to_le_bits()
+                        .iter()
+                        .by_val()
+                        .take(L_ORCHARD_MERKLE),
+                )
+                .chain(
+                    pair.right
+                        .to_le_bits()
+                        .iter()
+                        .by_val()
+                        .take(L_ORCHARD_MERKLE),
+                ),
+        )
+        .unwrap()
+}
+
+/// Generators for property testing.
+#[cfg(any(test, feature = "test-dependencies"))]
+pub mod testing {
+    use lazy_static::lazy_static;
+    use std::convert::TryInto;
+    use std::iter;
+
+    use crate::{
+        constants::MERKLE_DEPTH_ORCHARD,
+        note::{commitment::ExtractedNoteCommitment, testing::arb_note, Note},
+        value::{testing::arb_positive_note_value, MAX_NOTE_VALUE},
+    };
+    use pasta_curves::{arithmetic::FieldExt, pallas};
+
+    use proptest::collection::vec;
+    use proptest::prelude::*;
+
+    use super::{hash_layer, Anchor, MerklePath, Pair};
+
+    // The uncommitted leaf is defined as pallas::Base(2).
+    // <https://zips.z.cash/protocol/protocol.pdf#thmuncommittedorchard>
+    lazy_static! {
+        static ref EMPTY_ROOTS: Vec<pallas::Base> = {
+            iter::empty()
+                .chain(Some(pallas::Base::from_u64(2)))
+                .chain((0..MERKLE_DEPTH_ORCHARD).scan(
+                    pallas::Base::from_u64(2),
+                    |state, l_star| {
+                        *state = hash_layer(
+                            l_star,
+                            Pair {
+                                left: *state,
+                                right: *state,
+                            },
+                        );
+                        Some(*state)
+                    },
+                ))
+                .collect()
+        };
+    }
+
+    #[test]
+    fn test_vectors() {
+        let tv_empty_roots = crate::test_vectors::commitment_tree::test_vectors().empty_roots;
+
+        for (height, root) in EMPTY_ROOTS.iter().enumerate() {
+            assert_eq!(tv_empty_roots[height], root.to_bytes());
+        }
+    }
+
+    prop_compose! {
+        /// Generates an arbitrary Merkle tree of with `n_notes` nonempty leaves.
+        pub fn arb_tree(n_notes: usize)
+        (
+            // generate note values that we're certain won't exceed MAX_NOTE_VALUE in total
+            notes in vec(
+                arb_positive_note_value(MAX_NOTE_VALUE / n_notes as u64).prop_flat_map(arb_note),
+                n_notes
+            ),
+        )
+        -> (Vec<(Note, MerklePath)>, Anchor) {
+            // Inefficient algorithm to build a perfect subtree containing all notes.
+            let perfect_subtree_depth = (n_notes as f64).log2().ceil() as usize;
```

Comment:
Does the `as` cast work properly here, or should `into()` be used?

Reply:
Per https://doc.rust-lang.org/nomicon/casts.html:
> - casting from an integer to float will produce the floating point representation of the integer, rounded if necessary (rounding to nearest, ties to even)

I think this is fine; we cannot have more notes than `u32` given the depth of the tree, so that will fit into an `f64`.

Reply:
However, we could also use:
```suggestion
            let perfect_subtree_depth = n_notes.next_power_of_two();
```

Reply:
That's not the same thing; you want the log2 of the number of notes.

Reply:
Aha yep. Whoops!